tests-web/www/js/controls.js

```diff
@@ -0,0 +1,197 @@
+      layer.enabled = enabled;
+    }
+  }
+
+  renderControls() {
+    var members = [
+      makeCheckbox(
+        this.name,
+        `<u>${this.name}</u>`,
+        this.enabled,
+        (checked) => {
+          this.setEnabled(checked);
+
+          // Rerender
+          document.getElementById(this.name).replaceWith(this.renderControls());
+        }
+      ),
+    ];
+    for (const layer of this.layers) {
+      members.push(
+        makeCheckbox(
+          this.name + "_" + layer.name,
+          layer.name,
+          layer.enabled,
+          (checked) => {
+            layer.enabled = checked;
+            if (checked) {
+              this.map.addLayer(layer.layer);
+            } else {
+              this.map.removeLayer(layer.layer);
+            }
+          }
+        )
+      );
+
+      // This is maybe an odd time to sync this state
+      if (layer.enabled) {
+        this.map.addLayer(layer.layer);
+      } else {
+        this.map.removeLayer(layer.layer);
+      }
+    }
+    var div = makeDiv(members);
+    div.id = this.name;
+    return div;
+  }
+}
+
+// Manages a list of groups
+const LayerControl = L.Control.extend({
+  options: {
+    position: "bottomleft",
+  },
+  onAdd: function (map) {
+    return this.renderControls();
+  },
+
+  renderControls: function () {
+    var members = [];
+    for (const group of this.options.groups) {
+      members.push(group.renderControls());
+    }
+    var group = makeDiv(members);
+    group.style = "background: black; padding: 10px;";
+    L.DomEvent.disableClickPropagation(group);
+    return group;
+  },
+
+  getLayer: function (groupName, layerName) {
+    for (const group of this.options.groups) {
+      if (group.name == groupName) {
+        for (const layer of group.layers) {
+          if (layer.name == layerName) {
+            return layer.layer;
+          }
+        }
+      }
+    }
+
+    throw `Can't find group ${groupName} with layer ${layerName}`;
+  },
+
+  removeGroups: function (predicate) {
+    var keep = [];
+    for (const group of this.options.groups) {
+      if (predicate(group.name)) {
+        for (const layer of group.layers) {
+          group.map.removeLayer(layer.layer);
+        }
+      } else {
+        keep.push(group);
+      }
+    }
+    this.options.groups = keep;
+
+    L.DomUtil.empty(this.getContainer());
+    this.getContainer().appendChild(this.renderControls());
+  },
+
+  addGroup: function (group) {
+    this.options.groups.push(group);
+
+    L.DomUtil.empty(this.getContainer());
+    this.getContainer().appendChild(this.renderControls());
+  },
+});
+
+export const makeLayerControl = function (app) {
+  return new LayerControl({ app: app, groups: [] });
+};
+
+// Helpers
+
+function makeCheckbox(id, label, enabled, callback) {
+  var checkbox = L.DomUtil.create("input");
+  checkbox.id = id;
+  checkbox.type = "checkbox";
+  checkbox.checked = enabled;
+  checkbox.onclick = () => {
```

Comment:
Leaflet docs say to use these `DomEvent` helpers and do all this cleanup. I'm not sure why. We might be leaking DOM elements or something right now, but the end-user behavior seems correct to me